src/mbgl/renderer/image_manager.cpp

```diff
@@ -169,35 +170,53 @@ void ImageManager::reduceMemoryUseIfCacheSizeExceedsLimit() {
 }
 
 void ImageManager::checkMissingAndNotify(ImageRequestor& requestor, const ImageRequestPair& pair) {
-    std::vector<std::string> missingImages;
-    missingImages.reserve(pair.first.size());
+    ImageDependencies missingDependencies;
+
     for (const auto& dependency : pair.first) {
         if (images.find(dependency.first) == images.end()) {
-            missingImages.push_back(dependency.first);
+            missingDependencies.emplace(dependency);
         }
     }
 
-    if (!missingImages.empty()) {
+    if (!missingDependencies.empty()) {
         ImageRequestor* requestorPtr = &requestor;
+        assert(!missingImageRequestors.count(requestorPtr));
+        missingImageRequestors.emplace(requestorPtr, pair);
 
-        auto emplaced = missingImageRequestors.emplace(requestorPtr, MissingImageRequestPair{pair, {}});
-        assert(emplaced.second);
-
-        for (const auto& missingImage : missingImages) {
+        for (const auto& dependency : missingDependencies) {
+            const std::string& missingImage = dependency.first;
             assert(observer != nullptr);
-            requestedImages[missingImage].emplace(&requestor);
-            auto callback =
-                std::make_unique<ActorCallback>(*Scheduler::GetCurrent(), [this, requestorPtr, missingImage] {
-                    auto requestorIt = missingImageRequestors.find(requestorPtr);
-                    if (requestorIt != missingImageRequestors.end()) {
-                        assert(requestorIt->second.callbacks.find(missingImage) != requestorIt->second.callbacks.end());
-                        requestorIt->second.callbacks.erase(missingImage);
-                    }
-                });
-
-            auto actorRef = callback->self();
-            emplaced.first->second.callbacks.emplace(missingImage, std::move(callback));
-            observer->onStyleImageMissing(missingImage, [actorRef] { actorRef.invoke(&Callback::operator()); });
+
+            auto existingRequestorsIt = requestedImages.find(missingImage);
+            if (existingRequestorsIt != requestedImages.end()) { // Already asked client about this image.
+                std::set<ImageRequestor*>& existingRequestors = existingRequestorsIt->second;
+                if (!existingRequestors.empty() &&
+                    (*existingRequestors.begin())
+                        ->hasPendingRequest(missingImage)) { // Still waiting for the client response for this image.
+                    requestorPtr->addPendingRequest(missingImage);
+                    existingRequestors.emplace(requestorPtr);
+                    continue;
+                }
+                // The request for this image has been already delivered
+                // to the client, so we do not treat it as pending.
+                existingRequestors.emplace(requestorPtr);
```

Comment:
If client was asked to provide missing image, why do we ask again?

Reply:
the `onStyleImageMissing` handler in the `Map::Impl` re-invokes the rendering pipeline and we depend on this logic, otherwise map is gonna hang (see mapbox/api-gl#1404)